plugins/wu_palmer.py

```diff
@@ -267,10 +277,154 @@ def get_requirements(self) -> str:
+            continue
+
+        ancestors = [node_id]
+        visited = {node_id}
+        current_node_id = node_id
+
+        for _ in range(len(edges)):
+            current_node_id = edges.get(current_node_id, None)
+
+            if current_node_id is None:
+                break
+
+            if current_node_id in visited:
+                TASK_LOGGER.warn(
+                    f"cycle detected, node {current_node_id} already visited"
+                )
+                return None
+
+            ancestors.append(current_node_id)
+            visited.add(current_node_id)
+
+        if root_node is not None:
+            if root_node == ancestors[-1]:
+                root_node = None
+
+            if not root_node:
+                root_node = ancestors[-1]
+
+        nodes[node_id] = tuple(ancestors)[::-1]
+
+    return nodes
+
+
+class WuPalmerCache:
+    def __init__(self, taxonomies: Dict[str, Dict], root_has_meaning_in_taxonomy: bool):
+        self._taxonomies = taxonomies
+        self._node_path_cache: Dict[str, Dict[str, Tuple[str, ...]]] = {}
+        self._root_has_meaning_in_taxonomy = root_has_meaning_in_taxonomy
+
+    @lru_cache
+    def calculate_similarity(self, tax_name: str, node_a: str, node_b: str) -> float:
+        if tax_name not in self._node_path_cache:
+            self._node_path_cache[tax_name] = load_taxonomy_as_node_paths(
+                self._taxonomies[tax_name]
+            )
 
+        node_paths = self._node_path_cache[tax_name]
+
+        if node_a not in node_paths:
+            raise RuntimeError(f"node {node_a} not in node paths of taxonomy {tax_name}")
+
+        if node_b not in node_paths:
+            raise RuntimeError(f"node {node_b} not in node paths of taxonomy {tax_name}")
+
+        ancestors_a = node_paths[node_a]
+        ancestors_b = node_paths[node_b]
+
+        common_ancestor_depth = 0 if self._root_has_meaning_in_taxonomy else -1
+
+        for a, b in zip(ancestors_a, ancestors_b):
+            if a != b:
+                break
+
+            common_ancestor_depth += 1
+
+        if self._root_has_meaning_in_taxonomy:
+            denominator = len(ancestors_a) + len(ancestors_b)
+        else:
+            denominator = len(ancestors_a) + len(ancestors_b) - 2
+
+        if denominator == 0:
+            return 1
+        else:
+            wu_palmer_similarity = (2 * common_ancestor_depth) / denominator
+
+            return wu_palmer_similarity
+
+
+def add_similarities_for_entities(
+    similarities: Dict[Tuple[any, any], Dict],
+    entity1: Dict[str, any],
+    entity2: Dict[str, any],
+    attribute: str,
+    entities_metadata: Dict,
+    wu_palmer_cache: WuPalmerCache,
+):
+    if attribute not in entity1 or attribute not in entity2:
+        return
+
+    values1 = entity1[attribute]
+    values2 = entity2[attribute]
+
+    # extract taxonomy name from refTarget
+    tax_name: str = entities_metadata[attribute]["refTarget"].split(":")[1][:-5]
```

Comment:
Use pathlib to strip file name suffixes reliably instead of discarding the last n characters!